File: external/examples/GC-FID/GC_FID_optimization.py
import os
import matlab.engine
import GC_FID
from mesa.optimization import pipeline_iterator, plot_heatmaps
from mesa import optimization

if __name__ == "__main__":
    compute = False
    plot = True

    if compute:
        parameters = {
            "window_length_fraction": [1, 0.5, 0.3, 0.25, 0.20, 0.10, 0.05, 0.025, 0.01, 0.005, 0.0025],
            "hop_fraction": [1, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1],

            "window_function": ["boxcar", "hann", "blackman"],
        }

        gc_fid_metrics = [
            "Mean_SS", "Mean_perc_SS",
            "SS_time", "SS_treatment", "SS_sex", "SS_order", "SS_residuals",
            "F_time", "F_treatment", "F_sex", "F_order", "F_residuals",
            "p_time", "p_treatment", "p_sex", "p_order"
        ]

        pipeline = optimization.MatlabPipelineWrapper(
            matlab_script_folder="external/examples/GC-FID",
            matlab_script_name="GC_FID_optimization",
            output_metrics = gc_fid_metrics,
            python_pipeline_hook = GC_FID.main,
            custom_added_path="external/examples/GC-FID/MEDA"
        )

        try:
            pipeline.start_session()
            pipeline_iterator(parameters, pipeline, "external/examples/GC-FID/optimization.csv")
        finally:
            pipeline.stop_session()
    if plot:
        # plot_heatmaps("external/examples/GC-FID/optimization.csv",
        #             'hop_fraction',
        #             'window_length_fraction',
        #             'window_function',
        #             ['Mean_SS', 'Mean_perc_SS'],
        #             cmap = "ocean_r")
        # plot_heatmaps("external/examples/GC-FID/optimization.csv",
        #             'hop_fraction',
        #             'window_length_fraction',
        #             'window_function',
        #             ['SS_time', 'SS_treatment', 'SS_sex', 'SS_order'],
        #             vmin = 0, vmax = 30,
        #             cmap = "ocean_r")
        fig = plot_heatmaps("external/examples/GC-FID/optimization.csv",
                    'hop_fraction',
                    'window_length_fraction',
                    'window_function',
                    # ['F_time', 'F_treatment', 'F_sex', 'F_order'],
                    ['F_time'],
                    vmin = 0, vmax = 30,
                    cmap = "ocean_r")
        # plot_heatmaps("external/examples/GC-FID/optimization.csv",
        #             'hop_fraction',
        #             'window_length_fraction',
        #             'window_function',
        #             ['SS_residuals'],
        #             vmin = 0, vmax = 100,
        #             cmap="ocean_r")
        # plot_heatmaps("external/examples/GC-FID/optimization.csv",
        #             'hop_fraction',
        #             'window_length_fraction',
        #             'window_function',
        #             ['p_time', 'p_treatment', 'p_sex', 'p_order'], vmin = 0, vmax = 0.10,
        #             cmap="YlGnBu_r")

File: external/examples/GC-FID/GC_FID_optimization.m
function [Mean_SS, Mean_perc_SS, SS_time, SS_treatment, SS_sex, SS_null, ...
    SS_order, SS_residuals,F_time, F_treatment, F_sex, F_order, F_null,...
    F_residuals, p_time, p_treatment, p_sex, p_order, p_null, ...
    dist_time, dist_treatment, dist_sex, dist_order, dist_null] = GC_FID_optimization()
%% load data
close all;
load external/examples/GC-FID/GC-FID.mat

%% Preprocess
X = preprocess2D(data, 'Preprocessing', 0);
%% ASCA 
%% parglm
F = F_data; 

[T, parglmo] = parglm(X, F, 'Preprocessing', 2, 'Model', 'linear');
factor_names = ["Time", "Treatment", "Sex", "Order"];
for i = 1:size(factor_names,2)
    T{i+1,1} = {factor_names(i)};
end
%% build model
ascao = asca(parglmo);
ascao.nFactors = ascao.n_factors;
%%
SS_time      = T{2,3};
SS_treatment = T{3,3};
SS_sex       = T{4,3};
SS_order     = T{5,3};
SS_null      = T{6,3};
SS_residuals = T{7,3};

F_time      = T{2,6};
F_treatment = T{3,6};
F_sex       = T{4,6};
F_order     = T{5,6};
F_null      = T{6,6};
F_residuals = T{7,6};

p_time      = T{2,7};
p_treatment = T{3,7};
p_sex       = T{4,7};
p_order     = T{5,7};
p_null      = T{6,7};

Mean_SS = T{1,2};
Mean_perc_SS = T{1,3};


% distances = zeros(ascao.nFactors, 1);
% for factor_id = 1:ascao.nFactors
%     classes = F(:,factor_id);
%     levels = unique(classes);
    
%     pcs = max(ascao.factors{factor_id}.lvs);
%     means = zeros(length(levels), pcs);
%     for level_id = 1:size(levels,1)
%         idx = (classes == levels(level_id));
%         values = ascao.factors{factor_id}.scoresV(idx,:);
%         means(level_id,:) = mean(values);
%     end

%     distance = 0;
%     num_groups = size(means, 1);
    
%     % Doble bucle para comparar cada grupo con los siguientes
%     for i = 1:num_groups - 1
%         for j = i + 1:num_groups
%             % Calcula la distancia euclidiana entre la media del grupo i y el grupo j
%             diff = means(i, :) - means(j, :);
%             dist_ij = sqrt(sum(abs(diff).^2)); 
            
%             % Acumula la distancia
%             distance = distance + dist_ij;
%         end
%     end
%     distances(factor_id) = distance;

% end
% % Normalization according to number of variables
% M = size(X, 2);
% dist_time = distances(1)/M;
% dist_treatment = distances(2)/M;
% dist_sex = distances(3)/M;
% dist_order = distances(4)/M;
% dist_null = distances(5)/M;

%% Shilouette
distances = zeros(ascao.nFactors, 1);

for factor_id = 1:ascao.nFactors
    classes = F(:, factor_id);
    pcs = max(ascao.factors{factor_id}.lvs);
    scores = ascao.factors{factor_id}.scoresV(:, 1:pcs);
    scores = abs(scores);

    sil_values = silhouette(scores, classes, 'Euclidean');
    
    distances(factor_id) = mean(sil_values);
end

% Silhouette is already normalized by definition between -1 and 1, 
% so dividing by M (number of variables) is likely no longer necessary.
dist_time      = distances(1);
dist_treatment = distances(2);
dist_sex       = distances(3);
dist_order     = distances(4);
dist_null      = distances(5);


end